src/reader.rs

```diff
@@ -325,80 +341,32 @@ impl<B: BufRead> Reader<B> {
     ///
     /// Note: depending on the start of the Event, we may need to read more
     /// data, thus we need a mutable buffer
-    fn read_bang<'a, 'b>(
-        &'a mut self,
-        buf_start: usize,
-        buf: &'b mut Vec<u8>,
-    ) -> Result<Event<'b>> {
-        if buf[buf_start..].starts_with(b"!--") {
-            while buf.len() < buf_start + 5 || !buf.ends_with(b"--") {
-                buf.push(b'>');
-                match read_until(&mut self.reader, b'>', buf, &mut self.buf_position) {
-                    Ok(0) => {
-                        self.buf_position -= buf.len() - buf_start;
-                        return Err(Error::UnexpectedEof("Comment".to_string()));
-                    }
-                    Ok(_) => (),
-                    Err(e) => return Err(e.into()),
-                }
-            }
-            let len = buf.len();
+    fn read_bang<'a, 'b>(&'a mut self, buf: &'b [u8]) -> Result<Event<'b>> {
```

Comment:
You're checking the starts twice between this `read_bang` and `read_bang_element` fn. We could probably find a way to do it only once (either merge the fn but you may need to repeat it for each trait, or have a proper `read_comment`, `read_cdata` and `read_doctype`).